### ISSUE TITLE
Memory management bug in IfcOpenShell Python

### ISSUE THREAD

**Issue by dtip** [IfcOpenShell/IfcOpenShell#1124]
Hi folks,

Thanks for your work on IfcOpenShell. I've been trying it out and have run into a strange bug which I can't figure out.

For any IFC file I read, for every product within the file, the first two coordinates of the first vertex are some extremely small number. I believe they should be zero. The first few values in the faces array are invalid, but slightly differently invalid depending on the build:

When building from GitHub: The first and third values for the first triangle face are some extremely large (clearly invalid) integer, typically on the order of 10^7.
When using the Conda build: The first and third values in the faces array are either extremely large positive integers (10^7) or extremely large negative integers (-10^9). The second and fourth values in the faces array are a large integer (10^4) and are equal to each other.

These values are non-deterministic and change each time the Python script runs.

Any idea what's going on here?

Happy to help debug if you can provide instructions.

### Code to reproduce

```python
import ifcopenshell
import ifcopenshell.geom

ifc_file = ifcopenshell.open("some_file.ifc")
walls = ifc_file.by_type("IfcWall") # Same problem if this is changed to IfcProducts
for wall in walls:
    shape = ifcopenshell.geom.create_shape(settings, wall).geometry
    verts = shape.verts
    faces = shape.faces    
    print(verts) # (1.8579311e-316, 1.61328876e-316, 0.0, ... the rest is correct)
    print(faces) # (41427728, 0, 32653328, ... the rest is correct)
```

### Environment
Ubuntu 20.04

- Conda env (fresh environment)
  - Python 3.7.8
  - oce v0.18.3 build 3
  - IfcOpenShell v0.6.0 build py37h0d10d40_3
- System env
  - Python 3.8.5
  - required libs installed through `apt`
  - built IfcOpenShell from source using the latest v0.6.0 branch

**Comment by aothms** [comment]
Quick question. Does it help if you change to:

```
x = ifcopenshell.geom.create_shape(settings, wall)
shape = x.geometry
```

Memory management in IfcOpenShell python has a couple of weird quirks and this might be one of them. With the default settings, the geometry is owned by the "Element" (as it's called in C++) and by not keeping a Python reference to it. It get's freed, freeing also the geometry.

Easiest way to fix it is to make sure somewhere in Python that the geometry carries a backreference to the Element.

**Comment by dtip** [comment]
Thanks for the speedy reply. That fixes it! Is there a way to fix this quirk or to warn users about it?

**Comment by aothms** [comment]
It's essentially this problem https://stackoverflow.com/questions/15983008/prevent-nested-c-struct-from-being-deleted-when-parent-is-gcd in which the answer gives very little solutions.

Like I said we need to establish a backreference to the parent element from the geometry. One thing you could experiment with (if you're interested) to prevent the garbage collection to kick in:

~~~
class geometry_element:
    def __init__(self, elem_from_cpp):
        self.elem_from_cpp = elem_from_cpp
        self.geometry = elem_from_cpp.geometry
        # Make sure geometry references parent object to prevent gc
        self.geometry.reference = self
    def __getattr__(self, k): return getattr(self.elem_from_cpp, k)
~~~

And then in the wrap_shape_creation() function make sure an element of this kind is returned.

Or as you suggest feel free to add line somewhere in the docs. Surely that would help too :)

**Comment by dtip** [comment]
@Moult thanks for updating the documentation. I'd like to suggest that a documentation change doesn't fix the problem. This issue should remain open so that either someone can fix it in the future or it'll serve as an extra signpost for people who have this problem.

Please consider re-opening.

**Comment by Moult** [comment]
Yes, you're right, perhaps the Python wrapper could also provide a fix.

**Comment by arunkvijay** [comment]
I'd like to report that I got a similar error while using `tree.select_ray` and querying the `r.position` from results. The XYZ positions were in the order of e-316 (like the above example). These errors happened while using the scripting interface in blenderbim and happens at random and goes away once I restart blenderbim.
